<a href="https://colab.research.google.com/github/JulioHenri/Tecnicas-de-amostragem/blob/main/tec_amostragem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importando bibliotecas

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split 

# Carregando e lendo o banco de dados

In [3]:
dados = pd.read_csv('HR_Engagement_Sat_Sales_UpdatedV4.0.csv') #carregando arquivo csv
dados.head() #lendo os 5 primeiros registros do dataset

ID  ... Sensor_Proximity(1-highest/10-lowest)
0   1  ...                                     9
1   2  ...                                     8
2   3  ...                                     8
3   4  ...                                     6
4   5  ...                                    10

[5 rows x 62 columns]

###Excluindo variáveis não significantes para o estudo

variáveis com média e/ou variância nula, muito pequena ou com muitos dados faltantes.

In [4]:
dados = dados.drop(['ID','Name','Rising_Star','Critical','Trending Perf','Talent_Level','Validated_Talent_Level','EMP_Sat_OnPrem_1','CSR Factor','sales','Women_Leave',
           'Men_Leave','EMP_Sat_OnPrem_2','EMP_Sat_OnPrem_3','EMP_Sat_OnPrem_4','EMP_Sat_OnPrem_5', 'EMP_Sat_Remote_3',	'EMP_Sat_Remote_4', 'EMP_Engagement_1',
           'EMP_Engagement_2','EMP_Engagement_3','EMP_Engagement_4','EMP_Engagement_5'], axis = 1)

In [5]:
dados.describe()

Will_Relocate  ...  Sensor_Proximity(1-highest/10-lowest)
count   14999.000000  ...                           14999.000000
mean        0.499833  ...                               6.399360
std         0.500017  ...                               2.588091
min         0.000000  ...                               1.000000
25%         0.000000  ...                               5.000000
50%         0.000000  ...                               7.000000
75%         1.000000  ...                               8.000000
max         1.000000  ...                              10.000000

[8 rows x 34 columns]

In [6]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 39 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   Department                             14999 non-null  object 
 1   GEO                                    14999 non-null  object 
 2   Role                                   14999 non-null  object 
 3   Will_Relocate                          14999 non-null  int64  
 4   Percent_Remote                         14999 non-null  float64
 5   EMP_Sat_Remote_1                       14999 non-null  int64  
 6   EMP_Sat_Remote_2                       14999 non-null  int64  
 7   EMP_Sat_Remote_5                       14999 non-null  int64  
 8   last_evaluation                        14999 non-null  float64
 9   number_project                         14999 non-null  int64  
 10  average_montly_hours                   14999 non-null  int64  
 11  ti

Podemos notar acima que há 14999 registros e 62 colunas. Também há várias atributos com valores faltantes que serão analisados posteriormente.

# Exercícios

Exercício 1.1

a. unidade de pesquisa: Se o empregado continua ou não na empresa

b. população: Funcionários de uma empresa

c. instrumento de coleta de dados: Sistema de cadastro da empresa

d. unidade respondente: Empresas na área de Recursos Humanos

e. possível sistema de referência: 

f. unidade amostral mais provável: Apenas funcionários com salário "baixo"

g. unidades amostrais alternativas: Funcionários do departamento de Finanças e  os que não foram promovidos nos últimos 5 anos

Exercício 1.5 

a. Árvores em uma floresta: Nesse caso, iria fazer a amostra aleatória sistemática pois ela é de rápida e fácil execução e dessa maneira poderíamos selecionar as diversas árvores e separa-las por características específicas e depois de ter feito isso, e depois escolher um número igual de árvores para cada grupo. Sendo assim, tentando ao máximo ter uma amostra que represente a população. Poderíamos fazer um estudo que prevê se uma árvore será desmatada ou não, baseado em suas características, tempo de vida e localidade em que se encontra.

b. Crianças abaixo de 5 anos e que tiveram sarampo:
Como a população em questão já é bem ninchada (específica), a amostragem aleatória simples é o mais indicado pela sua facilidade de aplicação. A variável a ser estudada poderia ser as sequelas que cada criança teve e a gravidade que a doença causou, para isso, iriamos ter que saber quanto tempo ele esteve no hospital, quanto tempo esteve com a doença, e características físicas e mentais de cada criança

c. Operários em indústrias têxteis: Amostragem estratificada é o melhor tipo de forma de obter uma amostra para esta população, pois separamos a população em grupos e subgrupos e após isso, podemos usar amostragem aleatória simples para cada grupo especificado. Através de entrevistas, poderíamos estudar a satisfação dos funcionários em relação ao trabalho e concluir cada avaliação dos empregados entre três classes (bom, médio ou ruim). Para obter essas informações, os questionários teriam informações pessoais de cada operário (idade, raça, gênero) e informações relacionadas ao trabalho como quanto tempo trabalha nessa indústria, se tem parentes trabalhando no mesmo setor e etc.



# Amostragem Simples

Simplesmente selecionar alguns valores do banco de dados de maneira aleatória. Nesse caso irei selecionar 40% dos meus registros.

In [7]:
amostra1 = dados.sample(frac=.4)
amostra1.head()

Department  ... Sensor_Proximity(1-highest/10-lowest)
12670               IT  ...                                     6
666         Operations  ...                                    10
904                 IT  ...                                     9
13726               IT  ...                                     7
5337   Human Resources  ...                                     9

[5 rows x 39 columns]

In [8]:
amostra1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6000 entries, 12670 to 10853
Data columns (total 39 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   Department                             6000 non-null   object 
 1   GEO                                    6000 non-null   object 
 2   Role                                   6000 non-null   object 
 3   Will_Relocate                          6000 non-null   int64  
 4   Percent_Remote                         6000 non-null   float64
 5   EMP_Sat_Remote_1                       6000 non-null   int64  
 6   EMP_Sat_Remote_2                       6000 non-null   int64  
 7   EMP_Sat_Remote_5                       6000 non-null   int64  
 8   last_evaluation                        6000 non-null   float64
 9   number_project                         6000 non-null   int64  
 10  average_montly_hours                   6000 non-null   int64  
 11 

In [83]:
print('a média é: {:.2f}'.format(amostra1['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância é: {:.2f}'.format(amostra1['Sensor_Heartbeat(Average/Min)'].std()))


a média é: 77.91
A variância é: 7.18


Obtendo a amostra, temos 6000 registros, selecionando 40% da população de maneira aleatória. 

# Amostragem Sistemática

Os elementos da população serão colocados em uma lista e cada enésimo elemento da lista é escolhido (sistematicamente) e irá compor a amostra.

In [53]:
seed = np.random.choice(10,1) #selecionando número aleatório de 1 a 10
seed

array([9])

In [54]:
indices = np.arange(0, len(dados), seed)
indices

array([    0,     9,    18, ..., 14976, 14985, 14994])

No primeiro parâmetro do código acima eu seleciono em qual número o array deve se iniciar, no segundo em que número ele deve parar e por último, o intervalo com que os números devem ser escolhidos.


**Obs1**: o método 'np.arange' retorna um array (lista) de valores.

**Obs2**: len(dados) retorna a quantidade total de registros do DataFrame **dados**.

In [55]:
amostra2 = dados.loc[indices]

No código acima eu localizei, dentro dos meus dados, somente os indíces que correspondem ao meu array 'indices', explicado anteriormente.

**OBS**: Índice de um DataFrame é o primeiro valor do registro da esquerda para direita, que não tem nenhuma label em cima.

In [56]:
amostra2.head()

Department  ... Sensor_Proximity(1-highest/10-lowest)
0        Operations  ...                                     9
9   Human Resources  ...                                     8
18       Operations  ...                                     6
27  Human Resources  ...                                     8
36       Operations  ...                                     6

[5 rows x 39 columns]

In [57]:
amostra2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1667 entries, 0 to 14994
Data columns (total 39 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   Department                             1667 non-null   object 
 1   GEO                                    1667 non-null   object 
 2   Role                                   1667 non-null   object 
 3   Will_Relocate                          1667 non-null   int64  
 4   Percent_Remote                         1667 non-null   float64
 5   EMP_Sat_Remote_1                       1667 non-null   int64  
 6   EMP_Sat_Remote_2                       1667 non-null   int64  
 7   EMP_Sat_Remote_5                       1667 non-null   int64  
 8   last_evaluation                        1667 non-null   float64
 9   number_project                         1667 non-null   int64  
 10  average_montly_hours                   1667 non-null   int64  
 11  tim

In [82]:
print('a média é: {:.2f}'.format(amostra2['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância é: {:.2f}'.format(amostra2['Sensor_Heartbeat(Average/Min)'].std()))

a média é: 77.84
A variância é: 7.25


Como podemos observar, a nossa amostra gerada varia a quantidade de registros de acordo com o número gerado aleatoriamente, selecionando assim, de maneira sistemática.

# Exemplo 2.1

In [14]:
#criando nova amostra com somente 3 registros.
nova_amostra = amostra2.sample(n=3)
nova_amostra

Department  ... Sensor_Proximity(1-highest/10-lowest)
13395  Human Resources  ...                                     7
9738        Operations  ...                                     6
2625   Human Resources  ...                                     7

[3 rows x 39 columns]

Para os dados descritos na nova amostra, os seguintes parâmetros populacionais podem ser definidos:

i. Departamento de cada funcionário

In [15]:
#Todos os departamentos dos registros obtidos na amostra
nova_amostra['Department'].unique()

array(['Human Resources', 'Operations'], dtype=object)

In [16]:
print('D = {} = Y'.format(nova_amostra['Department'].unique()))

D = ['Human Resources' 'Operations'] = Y


ii. Para o vetor (FiTi)'

In [17]:
# Obtendo os valores da última pontuação da avaliação desse funcionário
nova_amostra.last_evaluation

13395    0.94
9738     0.81
2625     0.53
Name: last_evaluation, dtype: float64

In [18]:
# Obtendo os valores dos números de projetos que o funcionário trabalhou
nova_amostra.number_project

13395    6
9738     3
2625     2
Name: number_project, dtype: int64

B = <br/> 
(0.99  0.94  0.61 <br/> 
3.00  7.00  5.00)<br/> 


Com relação à funções paramétricas populacionais, tem-se:


1. Tempo gasto em média na empresa.

In [19]:
# O método mean() devolve a média aritmética dos números dados a ele. 
nova_amostra['time_spend_company'].mean()

3.0

2. Média das variáveis 'last_evaluation' e 'number_project'.

In [20]:
print('média da última avaliação: {:.2f}'.format(nova_amostra['last_evaluation'].mean()))
print('média do número de projetos feitos pelos funcionários {}'.format(nova_amostra['number_project'].mean()))

média da última avaliação: 0.76
média do número de projetos feitos pelos funcionários 3.6666666666666665


3. Média de vezes que o funcionário visita o LinkedIn por nível de salário.

In [21]:
nova_amostra.groupby(by='salary').mean()['LinkedIn_Hits']

salary
low    11.0
Name: LinkedIn_Hits, dtype: float64

a. Total populacional


In [22]:
nova_amostra['time_spend_company'].sum()

9

b. Média populacional

In [23]:
nova_amostra['time_spend_company'].mean()

3.0

c. Variância populacional

In [24]:
nova_amostra['time_spend_company'].var()

1.0

d. Covariância populacional

In [25]:
import numpy as np

X = np.stack((nova_amostra['time_spend_company'], nova_amostra['number_project']), axis=0)
np.cov(X)

array([[1.        , 1.5       ],
       [1.5       , 4.33333333]])

e. Correlação populacional

In [26]:
#correlação
corr = nova_amostra.corr()

#valores absolutos
corr_target = abs(corr['time_spend_company'])

#Selecionando as features com maior correlação com a variável de interesse
relevant_features = corr_target[corr_target>0.8]
relevant_features

Will_Relocate                            0.866025
time_spend_company                       1.000000
Work_accident                            0.866025
left_Company                             0.866025
Emp_Work_Status_3                        0.866025
Emp_Collaborative_5                      0.866025
Sensor_Proximity(1-highest/10-lowest)    0.866025
Name: time_spend_company, dtype: float64

f. Razão populacional

In [27]:
x = nova_amostra['time_spend_company'].sum()
y = nova_amostra['number_project'].sum()
r_populacional = x/y
r_populacional

0.8181818181818182

g. Razão média populacional

In [28]:
r_populacional/3

0.27272727272727276

# Amostragem Estratificada

Consiste em dividir toda a população em estratos, de maneira que um indivíduo pode fazer parte apenas de um único estrato. Após tais camadas serem definidas, selecionam-se os indivíduos utilizando qualquer técnica de amostragem em cada um dos estratos de forma separada.

Esse método é muito utilizado em métodos de Aprendizagem de Máquina Supervisionada quando é separada uma amostra de teste e outra de treino. Nesse sentido, separamos a amostra abaixo de maneira estratificada para, quem sabe, futuros projetos de Machine Learning.


In [71]:
x = dados.drop('left_Company', axis=1) #excluindo somente a variável que nos diz se o funcionário em questão deixou ou não a empreas
y = dados['left_Company'] #selecionando somente a variável left_company (explicada acima) do dataset

#separando os dados em treino e teste
x_train, amostra3, y_train, y_test = train_test_split(x, y, test_size=0.3, stratify=y)


Agora, nosso banco de dados foi divido em 4 'amostras', onde em duas (y_train e y_test) estão somente registros da variável 'Department', e outras duas (x_train e x_test) que estão os restantes das variáveis. 

O que difere nessas amostras é a quantidade de registros, pois os que tem o final 'train', contêm 70% dos registros do banco de dados original, enquanto que os que tem final 'test' eu defini apenas 30% dos dados, como mostrado a seguir.

In [73]:
#primeiro número se refere ao número de registros e o segundo ao número de variáveis.
amostra3.shape

(4500, 38)

In [80]:
print('a média é: {:.2f}'.format(amostra3['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância é: {:.2f}'.format(amostra3['Sensor_Heartbeat(Average/Min)'].std()))

a média é: 78.01
A variância é: 7.15


# Amostragem por conglomerado

Consiste na seleção de pessoas em grupos. Esse método é muito usado em questionários por questão de praticidade. Diferentemente da amostragem estratificada, no qual todos os estratos devem fazer parte da amostra, a amostragem por conglomerado pode selecionar por amostragem.

No nosso caso queremos mapear somente os registros de funcionários do banco de dados que trabalham no setor de vendas (um público bem específico). Isso pode ser visto com os códigos abaixo:

In [34]:
#Agrupando por Departamento e o transformando em um dataframe com 30% dos registros
grupo = dados.groupby('Department').apply(pd.DataFrame.sample, frac=.3)

In [35]:
#dentro do grupo, selecionando somente o departamento de vendas 
amostra4 = grupo[grupo.Department=='Sales']

In [60]:
#mostrando a amostra selecionada
amostra4.head()

Department  ... Sensor_Proximity(1-highest/10-lowest)
Department                   ...                                      
Sales      9673       Sales  ...                                     2
           11995      Sales  ...                                     5
           14197      Sales  ...                                     3
           11341      Sales  ...                                     3
           4285       Sales  ...                                     7

[5 rows x 39 columns]

In [63]:
#750 registros com 39 colunas
amostra4.shape

(750, 39)

In [81]:
print('a média é: {:.2f}'.format(amostra4['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância é: {:.2f}'.format(amostra4['Sensor_Heartbeat(Average/Min)'].std()))

a média é: 77.94
A variância é: 7.37


# Comparando o valor das médias e das variâncias das amostras e da população.

In [86]:
print('A média da amostra 1 é: {:.2f}'.format(amostra1['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância da amostra 1 é: {:.2f}'.format(amostra1['Sensor_Heartbeat(Average/Min)'].std()))

print('A média da amostra 2 é: {:.2f}'.format(amostra2['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância da amostra 2 é: {:.2f}'.format(amostra2['Sensor_Heartbeat(Average/Min)'].std()))

print('A média da amostra 3 é: {:.2f}'.format(amostra3['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância da amostra 3 é: {:.2f}'.format(amostra3['Sensor_Heartbeat(Average/Min)'].std()))

print('A média da amostra 3 é: {:.2f}'.format(amostra4['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância da amostra 3 é: {:.2f}'.format(amostra4['Sensor_Heartbeat(Average/Min)'].std()))

print('A média da população é: {:.2f}'.format(dados['Sensor_Heartbeat(Average/Min)'].mean()))
print('A variância da população é: {:.2f}'.format(dados['Sensor_Heartbeat(Average/Min)'].std()))


A média da amostra 1 é: 77.91
A variância da amostra 1 é: 7.18
A média da amostra 2 é: 77.84
A variância da amostra 2 é: 7.25
A média da amostra 3 é: 78.01
A variância da amostra 3 é: 7.15
A média da amostra 3 é: 77.94
A variância da amostra 3 é: 7.37
A média da população é: 77.97
A variância da população é: 7.25
